### ISSUE TITLE
What is the intended way of substituting returning objects on method calls?

### ISSUE THREAD

**Issue by gpr-indevelopment** [javaparser/javaparser#2373]
I have been struggling to figure out how to properly substitute some methods in a project I am currently running. What is the proper way of doing this, making the best use of this amazing lib?

I have the following test case:
```
import net.weg.eng.service.object.ComponentBase;
import net.weg.parserpoc.sampleobjects.*; //ObjectA is inside.
public class TestClass {

    public void execute(ComponentBase componentBase){
        String testString = componentBase.get("TEST_CHAR_1");
    }
}
```

The intended conversion for the execute method would be:

```
public void execute(ObjectA componentBase){
        MaestroString testString = componentBase.getTest_char_1();
    }
```
I have been doing it this way: First I change the method declaration (ComponentBase -> ObjectA), then I change the method call expression (get("TEST_CHAR_1") -> getTest_char_1()) with the following lines of code:

```
declaration.getParameter(0).setType("ObjectA");
methodCallExpr.setName(fieldName);
methodCallExpr.setArguments(new NodeList<>());
```
Then I need to change the return type (previously String, to MaestroString). I noticed I cannot resolve the new getTest_char_1() method (is this intended?), even though I have ObjectA's .jar imported to the solver. I have tried some alternatives, but the return types are determined at runtime (I cannot hardcode it). I believe this implementation is something you are used to...

What is the proper way of changing variable declaration types, according to the method call expression return type?

**Comment by matozoid** [comment]
Am I correct that this is a question about the symbol solver? I see you're doing a lot of AST manipulations, but there seems to be nothing going wrong with that?

**Comment by gpr-indevelopment** [comment]
I am using AST manipulations to reach/change some nodes, like the method declaration and method call expression. 

But, I am also using symbol solver to resolve some method calls, so that I know from which object this call is coming from. With this information, I am able to alter the method call (via AST manipulation). My problem is related to the type of the return (String to MaestroString in my example). What is the most correct way of changing the return type, based on a new method (after the method call expression change)?

- I have changed the method `componentBase.get("TEST_CHAR_1")`, that has a String return type, to `componentBase.getTest_char_1()`, that must have a MaestroString return type (inferred at runtime). How can change the return type of the variable declaration based on the new method?

**Comment by jlerbsc** [comment]
You can modify the return type statically with MethodDeclaration.setType() but i'm not sure that answers your question.

    "but the return types are determined at runtime"

**Comment by matozoid** [comment]
I suspect some kind of caching going on here. What exactly happens here: " I cannot resolve the new getTest_char_1() method"? An exception?

@jlerbsc it would not be too hard to make a test case for this I think?

**Comment by gpr-indevelopment** [comment]
When I try to resolve the `getTest_char_1()` my class is in the following state (notice the return type is still String, as I am about to change it):

```
public void execute(ObjectA componentBase){
        **String** testString = componentBase.getTest_char_1();
    }
```
Then I use a visitor to retrieve variable declarations in this code. I am able to retrieve the `testString` variable declaration this way. With the following code I am able to reach the `getTest_char_1()` method call expression, that defines the new type of `testString`:

```
dto.getVariableDeclarations().get(0).getVariable(0).getInitializer().get().asMethodCallExpr()
```
If I try to resolve this method call expression (appending a `.resolve()`), I get the following exception:

```
UnsolvedSymbolException{context='null', name='We are unable to find the method declaration corresponding to componentBase.getTest_char_1()', cause='null'}
	at com.github.javaparser.symbolsolver.JavaSymbolSolver.resolveDeclaration(JavaSymbolSolver.java:146)
	at com.github.javaparser.ast.expr.MethodCallExpr.resolve(MethodCallExpr.java:313)
	at net.weg.parserpoc.parser.resolvers.VariableDeclarationResolver.resolve(VariableDeclarationResolver.java:29)
	at net.weg.parserpoc.parser.resolvers.ResolverFacade.resolveVariableDeclarations(ResolverFacade.java:28)
	at net.weg.parserpoc.parser.resolvers.ResolverFacade.resolve(ResolverFacade.java:16)
	at net.weg.parserpoc.MethodSubstitution.test(MethodSubstitution.java:34)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.springframework.test.context.junit4.statements.RunBeforeTestExecutionCallbacks.evaluate(RunBeforeTestExecutionCallbacks.java:74)
	at org.springframework.test.context.junit4.statements.RunAfterTestExecutionCallbacks.evaluate(RunAfterTestExecutionCallbacks.java:84)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:75)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:86)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:251)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```
The ideal situation here would be: Resolve the method after changing it to `getTest_char_1()`, retrieve the new reference, and use it to apply the new return type to the `testString` variable.

**Comment by matozoid** [comment]
I'm passing the case to @jlerbsc :-)

**Comment by jlerbsc** [comment]
you are too generous ;-)

**Comment by jlerbsc** [comment]
Is this approximately what you want? If not, please give us a complete test case, thanks.

    @Test
    void methodWithVarargsResolution() throws IOException {
        CombinedTypeSolver combined = new CombinedTypeSolver();
        combined.add(new JarTypeSolver("C:\\Users\\xxxx\\Downloads\\MaestroSolverConfig\\lib\\maestro.jar"));
        combined.add(new ReflectionTypeSolver());
        StaticJavaParser.getConfiguration().setSymbolResolver(new JavaSymbolSolver(combined));
        CompilationUnit cu = StaticJavaParser.parse(new File("C:\\Users\\xxxx\\Downloads\\MaestroSolverConfig\\B.java"));
        Optional<MethodDeclaration> md = cu.findFirst(MethodDeclaration.class);
        assertTrue(md.isPresent());

        System.out.println("before: " + md.toString());
        Optional<MethodCallExpr> methodCallExpr = md.get().findFirst(MethodCallExpr.class);
        Optional<VariableDeclarator> vd = md.get().findFirst(VariableDeclarator.class);
        String fieldName = methodCallExpr.get().getArguments().get(0).asStringLiteralExpr().asString();
    
        md.get().getParameter(0).setType("A");
        methodCallExpr.get().setName(fieldName);
        methodCallExpr.get().setArguments(new NodeList<>());
        vd.get().setType("MaestroString");
        
        System.out.println("after: " + md.toString());
    }

    before: Optional[public void execute(ComponentBase componentBase) {
        String testString = componentBase.get("TEST_CHAR_1");
    }]
    after: Optional[public void execute(A componentBase) {
        MaestroString testString = componentBase.TEST_CHAR_1();
    }]

with B.java

    public class B {
        public void execute(ComponentBase componentBase){
            String testString = componentBase.get("TEST_CHAR_1");
        }
    }

**Comment by gpr-indevelopment** [comment]
@jlerbsc almost correct. The problem I see is with this line: `vd.get().setType("MaestroString");`. There, you hard-code that the new return type is MaestroString. My situation is a little more complicated.

Since you changed the execute method's argument to be A, instead of ComponentBase, and the method to be TEST_CHAR_1 (called from A, instead of ComponentBase), I am not able to determine the MaestroString return, unless I resolve the new `componentBase.TEST_CHAR_1()` (with componentBase now referencing A).

The method `A.TEST_CHAR_1` may have pretty much any return type, and I need to determine it at runtime.

I need to resolve `A.TEST_CHAR_1`, so that I know from which class this method is coming from. Then I can access its original method declaration to retrieve what the correct return type should be.

Instead of hard-coding "MaestroString" I need to be able to look for it.

**Comment by jlerbsc** [comment]
Is it correct now?

    @Test
    void test() throws IOException {
        CombinedTypeSolver combined = new CombinedTypeSolver();
        combined.add(new JarTypeSolver("C:\\Users\\xxxx\\Downloads\\MaestroSolverConfig\\lib\\maestro.jar"));
        combined.add(new ReflectionTypeSolver());
        StaticJavaParser.getConfiguration().setSymbolResolver(new JavaSymbolSolver(combined));
        CompilationUnit cu = StaticJavaParser.parse(new File("C:\\Users\\xxxx\\Downloads\\MaestroSolverConfig\\B.java"));
        Optional<MethodDeclaration> md = cu.findFirst(MethodDeclaration.class);
        assertTrue(md.isPresent());
        
        Optional<MethodCallExpr> methodCallExpr = md.get().findFirst(MethodCallExpr.class);
        System.out.println("before: " + md.toString());
        Optional<VariableDeclarator> vd = md.get().findFirst(VariableDeclarator.class);
        String fieldName = methodCallExpr.get().getArguments().get(0).asStringLiteralExpr().asString();
        md.get().getParameter(0).setType("A");
        methodCallExpr.get().setName(fieldName);
        methodCallExpr.get().setArguments(new NodeList<>());
        // resolve the new componentBase.TEST_CHAR_1() (with componentBase now referencing A)         
        ResolvedMethodDeclaration resolved = methodCallExpr.get().resolve();
        // The method A.TEST_CHAR_1 may have pretty much any return type, and I need to determine it at runtime.
        ResolvedType type = resolved.getReturnType();
        vd.get().setType(type.describe());
        
        System.out.println("after: " + md.toString());
    }

    before: Optional[public void execute(ComponentBase componentBase) {
    String testString = componentBase.get("TEST_CHAR_1");
    }]
    after: Optional[public void execute(A componentBase) {
    java.lang.Object testString = componentBase.TEST_CHAR_1();
    }]

    public class A {
        public void execute(ComponentBase _this) {
		_this.getRoot().executeMethod("functionName","regexText","outpuText",1);
        }
	
        public Object TEST_CHAR_1 () {
		return null;
        }
    }

    public class B {
        public void execute(ComponentBase componentBase){
            String testString = componentBase.get("TEST_CHAR_1");
        }
    }

**Comment by gpr-indevelopment** [comment]
@jlerbsc perfect! That is exactly the case... I get the exception when I try to execute something similar to this in my code:

```
// resolve the new componentBase.TEST_CHAR_1() (with componentBase now referencing A)         
    ResolvedMethodDeclaration resolved = methodCallExpr.get().resolve();
```

**Comment by gpr-indevelopment** [comment]
@jlerbsc did you happen to take a look at it?

**Comment by jlerbsc** [comment]
If the test case is correct i cannot reproduce the issue! Could you modify this test class to generate your exception please.

**Comment by gpr-indevelopment** [comment]
So, was this test case able to infer the new return type of the changed method? Was it able to be resolved after the changes to the MethodCallExpression?

**Comment by gpr-indevelopment** [comment]
My intention with the question was to know what is the intended way of changing return types dynamically. Does this test reproduce how return type solving should be implemented?

**Comment by matozoid** [comment]
You're doing things right. @jlerbsc is trying to find out why it doesn't work, so he can fix the problem for you.

Alternatively, since this project runs on user contributions, you can consider diving in and suggesting a fix.

**Comment by gpr-indevelopment** [comment]
@matozoid sure! I am currently running an experimental project using JavaParser at work. If this project runs for longer, I will certainly dive into the code so I can make contributions. 

@jlerbsc I will take a look at your test case, to see if I can reproduce the error here. If not, I will try to write a test case for you. 

Thank you guys.

**Comment by matozoid** [comment]
No problem. A workaround might be to use `clone()`: if you start with a simple AST without the symbol solver connected and do your transformations on that, you can clone it and use that clone for solving.

**Comment by gpr-indevelopment** [comment]
@matozoid I ended up figuring it out: My program is generating classes dinamically, with getter methods. Then, I go over those classes to map all getter methods and their return types.

After that, whenever I need to change some method into one of those getter methods, I just looked for the mapped return type and substitute it on the VariableDeclarator parent object.

Can you think of a better way of doing so?

**Comment by matozoid** [comment]
Better ask @jlerbsc :-)

**Comment by jlerbsc** [comment]
No, ideally @gpr-indevelopment should provide a test case that reproduces the anomaly. Without that I will not spend more time trying to solve something that is not perfectly defined.

**Comment by matozoid** [comment]
Well, I guess that's as far as it goes then :-)

**Comment by jlerbsc** [comment]
Hi @gpr-indevelopment We have provided a unit test which reproduces the case. The return type of the method has been dynamically defined. This is one way of responding to this issue. We cannot do more about this problem. If you have other questions on this issue you can open a new issue or comment on this one.
